## VAR vs VARMAX Model Validation


In this notebook, we will validate which model performs better given computational constraints: a VAR model retrained after each prediction or a VARMAX retrained every 20 predictions. To do so, we will retrun the model and functions constructed in [Systems_Identification_Fitting](Systems_Identification_Fitting.ipynb).

In [615]:
# import libraries
import pandas as pd
import numpy as np
from scipy import stats
import math as m
from sklearn.metrics import mean_squared_error
import statsmodels.api as sm
from statsmodels.tsa.api import VAR, VARMAX
from sklearn.preprocessing import PowerTransformer
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

%matplotlib inline 


states = pd.read_csv('states.csv')
del states['Unnamed: 0']
states.head()

marketPriceEth  marketPriceUsd  block_number  debtAvailableToSettle  \
0        0.000858        3.051840    12388143.0                    0.0   
1        0.000858        3.051840    12387897.0                    0.0   
2        0.000869        3.060763    12387633.0                    0.0   
3        0.000870        3.051668    12387345.0                    0.0   
4        0.000875        3.043224    12387095.0                    0.0   

     globalDebt  globalDebtCeiling  systemSurplus  totalActiveSafeCount  \
0  2.632464e+07       1.157921e+32  230002.364951                 583.0   
1  2.632464e+07       1.157921e+32  230002.364951                 583.0   
2  2.629462e+07       1.157921e+32  229927.716817                 583.0   
3  2.629462e+07       1.157921e+32  229927.716817                 583.0   
4  2.629407e+07       1.157921e+32  229903.041060                 583.0   

   RedemptionRateAnnualizedRate  RedemptionRateHourlyRate  \
0                      0.978704                  0.999998   
1                      0.978704                  0.999998   
2                      0.978704                  0.999998   
3                      0.978704                  0.999998   
4                      0.978704                  0.999998   

   RedemptionRateEightHourlyRate  RedemptionPrice  EthInUniswap  RaiInUniswap  \
0                        0.99998         3.005443  11977.048864  1.389236e+07   
1                        0.99998         3.005450  12025.302340  1.383645e+07   
2                        0.99998         3.005465  12033.680708  1.382679e+07   
3                        0.99998         3.005465  12033.680708  1.382679e+07   
4                        0.99998         3.005472  12080.165571  1.377424e+07   

   RaiDrawnFromSAFEs  collateral           debt  ETH Price (OSM)  
0       2.601571e+07   1249.4226  293469.366076      3521.749280  
1       2.601571e+07   1249.4226  293469.366076      3484.967617  
2       2.598571e+07   1249.4226  293469.366076      3509.940000  
3       2.598571e+07   1249.4226  293469.366076      3438.463957  
4       2.598521e+07   1249.4226  293469.366076      3464.125574

In [616]:
# add additional state variables
states['RedemptionPriceinEth'] = states['RedemptionPrice'] / states['ETH Price (OSM)']
states['RedemptionPriceError'] = states['RedemptionPrice'] - states['marketPriceUsd']

In [617]:
# define constants (will come from cadCAD model but added here for calculations)
params = {
    'liquidation_ratio': 1.45,
    'debt_ceiling': 1e9,
    'uniswap_fee': 0.003,
    'arbitrageur_considers_liquidation_ratio': True,
}


<!-- ## Create Arbtrageur data vector $u^*$ -->

In [620]:
# subset state variables for arbitrageur vector
state_subset = states[['collateral','RaiDrawnFromSAFEs','RaiInUniswap','EthInUniswap']]

# map state data to vector fields 
state_subset.columns = ['Q','D','Rrai','Reth']

# alpha is the smoothing factor
local = state_subset.ewm(alpha=0.8).mean()
local

Q             D          Rrai          Reth
0     1249.422600  2.601571e+07  1.389236e+07  11977.048864
1     1249.422600  2.601571e+07  1.384577e+07  12017.260094
2     1249.422600  2.599152e+07  1.383046e+07  12030.502525
3     1249.422600  2.598687e+07  1.382752e+07  12033.049146
4     1249.422600  2.598554e+07  1.378488e+07  12070.754352
...           ...           ...           ...           ...
1942    18.086319  3.383805e+03  7.636681e+02      1.509233
1943    17.131152  3.115513e+03  7.942806e+02      1.436701
1944    16.940119  3.061855e+03  8.004031e+02      1.422195
1945    16.901912  3.051123e+03  8.016276e+02      1.419294
1946    16.894271  3.048977e+03  8.018725e+02      1.418714

[1947 rows x 4 columns]

In [621]:
# function to create coordinate transformations
def coordinate_transformations(params,df,Q,R_eth,R_rai,D,RedemptionPrice,EthPrice):
    '''
    Description:
    Function that takes in pandas dataframe and the names of columns
    
    Parameters:
    df: pandas dataframe containing states information
    Q: dataframe column name
    R_eth: dataframe column name
    R_rai: dataframe column name
    D: dataframe column name
    RedemptionPrice: dataframe column name
    EthPrice: dataframe column name

    Returns: Pandas dataframe with alpha, beta, gamma, delta transformed values
    
    Example:
    
    coordinate_transformations(params,states,'collateral','EthInUniswap','RaiInUniswap',
                           'RaiDrawnFromSAFEs','RedemptionPrice','ETH Price (OSM)')[['alpha','beta','gamma','delta']]
    '''
    
    # Calculate alpha
    d = df[D].diff()
    d.fillna(0,inplace=True)
    df['d'] = d
    
    df['alpha'] = df['d'] / params['debt_ceiling']

    # calculate beta
    df['C_o'] = (df[RedemptionPrice]/states[EthPrice]) * params['liquidation_ratio']

    q = df[Q].diff()
    q.fillna(0,inplace=True)
    df['q'] = q


    df['C_1'] = (df['C_o'] * df[D]) - df[Q]

    df['beta'] = (df['q'] - (df['C_o']*df['d']))/ df['C_1']

    # calculate gamma
    r = df[R_rai].diff()
    r.fillna(0,inplace=True)
    df['r'] = r

    df['gamma'] = df['r']/df[R_rai]

    # calculate delta
    z = df[R_eth].diff()
    z.fillna(0,inplace=True)
    df['z'] = z

    df['delta'] = df['z']/df[R_eth]
    
    return df

In [622]:
transformed = coordinate_transformations(params,states,'collateral','EthInUniswap','RaiInUniswap',
                           'RaiDrawnFromSAFEs','RedemptionPrice','ETH Price (OSM)')[['alpha','beta','gamma','delta']]

transformed

alpha      beta     gamma     delta
0     0.000000e+00  0.000000  0.000000  0.000000
1     0.000000e+00  0.000000 -0.004041  0.004013
2    -3.000000e-05  0.001201 -0.000699  0.000696
3     0.000000e+00  0.000000  0.000000  0.000000
4    -5.000000e-07  0.000020 -0.003815  0.003848
...            ...       ...       ...       ...
1942 -1.646075e-06  0.108212  0.249397 -0.333334
1943  0.000000e+00 -0.000000  0.000000  0.000000
1944  0.000000e+00 -0.000000  0.000000  0.000000
1945  0.000000e+00 -0.000000  0.000000  0.000000
1946  0.000000e+00 -0.000000  0.000000  0.000000

[1947 rows x 4 columns]

In [623]:
# add additional signals to arbitrageur state
local['RedemptionPrice'] = states['RedemptionPrice']
local['ETH Price (OSM)'] = states['ETH Price (OSM)']

local

Q             D          Rrai          Reth  RedemptionPrice  \
0     1249.422600  2.601571e+07  1.389236e+07  11977.048864         3.005443   
1     1249.422600  2.601571e+07  1.384577e+07  12017.260094         3.005450   
2     1249.422600  2.599152e+07  1.383046e+07  12030.502525         3.005465   
3     1249.422600  2.598687e+07  1.382752e+07  12033.049146         3.005465   
4     1249.422600  2.598554e+07  1.378488e+07  12070.754352         3.005472   
...           ...           ...           ...           ...              ...   
1942    18.086319  3.383805e+03  7.636681e+02      1.509233         3.140000   
1943    17.131152  3.115513e+03  7.942806e+02      1.436701         3.140000   
1944    16.940119  3.061855e+03  8.004031e+02      1.422195         3.140000   
1945    16.901912  3.051123e+03  8.016276e+02      1.419294         3.140000   
1946    16.894271  3.048977e+03  8.018725e+02      1.418714         3.140000   

      ETH Price (OSM)  
0         3521.749280  
1         3484.967617  
2         3509.940000  
3         3438.463957  
4         3464.125574  
...               ...  
1942      1825.887144  
1943      1793.770282  
1944      1803.891149  
1945      1803.891149  
1946      1805.792735  

[1947 rows x 6 columns]

In [624]:
transformed_arbitrageur = coordinate_transformations(params,local,'Q','Reth','Rrai',
                           'D','RedemptionPrice','ETH Price (OSM)')[['alpha','beta','gamma','delta']]

transformed_arbitrageur

alpha      beta     gamma     delta
0     0.000000e+00  0.000000  0.000000  0.000000
1     0.000000e+00  0.000000 -0.003365  0.003346
2    -2.419355e-05  0.000968 -0.001107  0.001101
3    -4.652605e-06  0.000186 -0.000213  0.000212
4    -1.323500e-06  0.000053 -0.003093  0.003124
...            ...       ...       ...       ...
1942 -1.341458e-06  0.148292  0.200431 -0.240292
1943 -2.682915e-07  0.029727  0.038541 -0.050485
1944 -5.365830e-08  0.006036  0.007649 -0.010200
1945 -1.073166e-08  0.001209  0.001528 -0.002044
1946 -2.146332e-09  0.000242  0.000305 -0.000409

[1947 rows x 4 columns]

In [625]:
def create_transformed_errors(transformed_states,transformed_arbitrageur):
    '''
    Description:
    Function for taking two pandas dataframes of transformed states and taking the difference
    to produce an error dataframe. 
    
    Parameters:
    transformed_states: pandas dataframe with alpha, beta, gamma, and delta features
    transformed_arbitrageur: pandas dataframe with alpha, beta, gamma, and delta features

    Returns:
    error pandas dataframe and transformation object
    
    '''
    alpha_diff = transformed_states['alpha'] - transformed_arbitrageur['alpha']
    beta_diff = transformed_states['beta'] - transformed_arbitrageur['beta']
    gamma_diff = transformed_states['gamma'] - transformed_arbitrageur['gamma']
    delta_diff = transformed_states['delta'] - transformed_arbitrageur['delta']


    e_u = pd.DataFrame(alpha_diff)
    e_u['beta'] = beta_diff
    e_u['gamma'] = gamma_diff
    e_u['delta'] = delta_diff

    e_u = e_u.astype(float)
    
    return e_u

e_u = create_transformed_errors(transformed,transformed_arbitrageur)
e_u.head()

alpha      beta     gamma     delta
0  0.000000e+00  0.000000  0.000000  0.000000
1  0.000000e+00  0.000000 -0.000676  0.000667
2 -5.806452e-06  0.000233  0.000408 -0.000404
3  4.652605e-06 -0.000186  0.000213 -0.000212
4  8.235004e-07 -0.000033 -0.000722  0.000724

In [626]:
def power_transformation(e_u):
    '''
    Definition:
    Function to perform a power transformation on the coordinate 
    transformed differenced data
    
    Parameters:
    e_u: Dataframe of coordinated transformed differenced data
    
    Required:
    import pandas as pd
    from sklearn.preprocessing import PowerTransformer
    
    Returns:
    Transformed dataframe and transformation object
    
    Example:
    transformed_df, pt = power_transformation(e_u)
    '''
    pt = PowerTransformer()
    yeo= pd.DataFrame(pt.fit_transform(e_u),columns=e_u.columns)

    return yeo, pt

In [627]:
e_u_yeo, pt = power_transformation(e_u)

<!-- ### Autogressive lag selection -->

In [628]:
# split data between train and test (in production deployment, can remove)
split_point = int(len(e_u) * .8)
train = e_u_yeo.iloc[0:split_point]
test = e_u_yeo.iloc[split_point:]


states_train = states.iloc[0:split_point]
states_test = states.iloc[split_point:]

In [629]:
def VARMAX_prediction(e_u,RedemptionPriceError,newRedemptionPriceError,steps=1,lag=1):
    '''
    Description:
    Function to train and forecast a VARMAX model one step into the future
    
    Parameters:
    e_u: errors pandas dataframe
    RedemptionPriceErrorPrevious: 1d Numpy array of RedemptionPriceError values
    newRedemptionPriceError: exogenous latest redemption price error signal - float
    steps: Number of forecast steps. Default is 1
    lag: number of autoregressive lags. Default is 1
    
    Returns:
    Numpy array of transformed state changes
    
    Example
    Y_pred = VARMAX_prediction(train,states_train['RedemptionPriceError'],
                  states_test['RedemptionPriceError'][0:5],steps=5,lag=1)
    '''
    # instantiate the VARMAX model object from statsmodels 
    model = VARMAX(endog=e_u.values,exog=RedemptionPriceError,
                   initialization='approximate_diffuse',measurement_error=True)

    # fit model with determined lag values
    results = model.fit(order=(lag,0))
    
    Y_pred = results.forecast(steps = steps, exog=newRedemptionPriceError)
    
    return Y_pred.values

In [630]:
def VAR_prediction(e_u,lag=1):
    '''
    Description:
    Function to train and forecast a VAR model one step into the future
    
    Parameters:
    e_u: errors pandas dataframe
    lag: number of autoregressive lags. Default is 1
    
    Returns:
    Numpy array of transformed state changes
    
    Example
    VAR_prediction(e_u,6)    
    '''
    # instantiate the VAR model object from statsmodels 
    model = VAR(e_u)

    # fit model with determined lag values
    results = model.fit(lag)
    
    lag_order = results.k_ar
    
    Y_pred = results.forecast(e_u.values[-lag_order:],1)

    
    return Y_pred[0]

In [631]:
varmax_predictions = VARMAX_prediction(train,states_train['RedemptionPriceError'],
                  states_test['RedemptionPriceError'].values[0:20],steps=20,lag=1)

In [632]:
var_predictions = []
train_var = train.copy()
for i in range(0,20): 
    var_predictions.append(VAR_prediction(train_var,15))
    train_var = train_var.append(test.iloc[i])

In [633]:
def invert_power_transformation(pt,prediction):
    '''
    Definition:
    Function to invert power transformation
    
    Parameters:
    pt: transformation object
    prediction: Numpy array of model state coordinate transformed percentage changes
    
    Required:
    import pandas as pd
    from sklearn.preprocessing import PowerTransformer
    
    Returns:
    inverted transformation numpy array
    
    Example:
    inverted_array = invert_power_transformation(pt,prediction)
    
    '''
    # transform back into coordinate system
    inverted = pt.inverse_transform(prediction)
    
    return inverted

In [634]:
# invert the power transformation
var_predictions = invert_power_transformation(pt,var_predictions)
varmax_predictions = invert_power_transformation(pt,varmax_predictions)

## Model Evaluation Comparision (in coordinate transformed and power transformed differenced space)  

In [635]:
var_predictions_df = pd.DataFrame(var_predictions,columns=['alpha','beta','gamma','delta'])
varmax_predictions_df = pd.DataFrame(varmax_predictions,columns=['alpha','beta','gamma','delta'])

## Alpha - Root Mean Square Error

In [636]:
var_alpha_rmse = m.sqrt(mean_squared_error(var_predictions_df['alpha'], test.head(20)['alpha']))
varmax_alpha_rmse = m.sqrt(mean_squared_error(varmax_predictions_df['alpha'], test.head(20)['alpha']))
if var_alpha_rmse >= varmax_alpha_rmse:
    print('VARMAX performs better by {}'.format(varmax_alpha_rmse-var_alpha_rmse))
else:
    print('VAR performs better by {}'.format(var_alpha_rmse-varmax_alpha_rmse))

VAR performs better by -2.3483390703571416e-05


## Beta - Root Mean Square Error

In [637]:
var_beta_rmse = m.sqrt(mean_squared_error(var_predictions_df['beta'], test.head(20)['beta']))
varmax_beta_rmse = m.sqrt(mean_squared_error(varmax_predictions_df['beta'], test.head(20)['beta']))
if var_beta_rmse >= varmax_beta_rmse:
    print('VARMAX performs better by {}'.format(varmax_beta_rmse-var_beta_rmse))
else:
    print('VAR performs better by {}'.format(var_beta_rmse-varmax_beta_rmse))

VAR performs better by -0.0004483374446575324


## Gamma - Root Mean Square Error

In [638]:
var_gamma_rmse = m.sqrt(mean_squared_error(var_predictions_df['gamma'], test.head(20)['gamma']))
varmax_gamma_rmse = m.sqrt(mean_squared_error(varmax_predictions_df['gamma'], test.head(20)['gamma']))
if var_gamma_rmse >= varmax_gamma_rmse:
    print('VARMAX performs better by {}'.format(varmax_gamma_rmse-var_gamma_rmse))
else:
    print('VAR performs better by {}'.format(var_gamma_rmse-varmax_gamma_rmse))

VAR performs better by -0.00028606775500304715


## Delta - Root Mean Square Error

In [639]:
var_delta_rmse = m.sqrt(mean_squared_error(var_predictions_df['delta'], test.head(20)['delta']))
varmax_delta_rmse = m.sqrt(mean_squared_error(varmax_predictions_df['delta'], test.head(20)['delta']))
if var_delta_rmse >= varmax_delta_rmse:
    print('VARMAX performs better by {}'.format(varmax_delta_rmse-var_delta_rmse))
else:
    print('VAR performs better by {}'.format(var_delta_rmse-varmax_delta_rmse))

VAR performs better by -0.00046840677073029663


## Model Evaluation Comparision as States

In [640]:
def inverse_transformation_and_state_update(Y_pred,previous_state,params):
    '''
    Description:
    Function to take system identification model prediction and invert transfrom and create new state
    
    Parameters:
    y_pred: numpy array of transformed state changes
    previous_state: pandas dataframe of previous state or 'current' state
    params: dictionary of system parameters
    
    Returns:
    pandas dataframe of new states 
    
    Example:
    inverse_transformation_and_state_update(Y_pred,previous_state,params)
    '''
    
    d_star = Y_pred[0] * params['debt_ceiling']
    
    q_star = previous_state['C_o'] * params['debt_ceiling'] * Y_pred[0] + previous_state['C_1'] * Y_pred[1]
    
    r_star = Y_pred[2] * previous_state['gamma'] * previous_state['RaiInUniswap']

    z_star = Y_pred[3] * previous_state['delta'] * previous_state['EthInUniswap']

    new_state = pd.DataFrame(previous_state[['collateral','EthInUniswap','RaiInUniswap','RaiDrawnFromSAFEs']].to_dict())
    new_state['Q'] = new_state['collateral'] + q_star
    new_state['D'] = new_state['RaiDrawnFromSAFEs'] + d_star
    new_state['R_Rai'] = new_state['RaiInUniswap'] + r_star
    new_state['R_Eth'] = new_state['EthInUniswap'] + z_star
    
    return new_state[['Q','D','R_Rai','R_Eth']]



In [641]:
VAR_new_states = []
index = -21
for i in range(0,20): 
    previous_state = states.iloc[train_var.index[index:index+1]]
    VAR_new_states.append(inverse_transformation_and_state_update(var_predictions[i],previous_state,params))
    index += 1
    
VAR_new_states = pd.concat(VAR_new_states)
VAR_new_states

Q             D         R_Rai         R_Eth
1556  6081.169224  3.371551e+07  2.971972e+07  61661.047608
1557  6110.991138  3.369966e+07  2.961955e+07  61701.017149
1558  6120.952802  3.413544e+07  3.018807e+07  60746.179731
1559  6087.703343  3.404582e+07  2.997961e+07  60989.915859
1560  6088.850114  3.411418e+07  2.998927e+07  60974.824543
1561  6085.478244  3.409787e+07  3.002661e+07  60899.884484
1562  6141.965543  3.397362e+07  2.977962e+07  60663.356884
1563  6091.936632  3.379863e+07  2.953994e+07  59906.282132
1564  6126.153590  3.353541e+07  2.943921e+07  60270.621867
1565  6131.086605  3.339909e+07  2.922044e+07  60086.935213
1566  6096.528881  3.338932e+07  2.919057e+07  60163.041470
1567  6102.356558  3.329873e+07  2.913818e+07  60212.233856
1568  6138.249445  3.331039e+07  2.902142e+07  60285.465969
1569  6116.777496  3.334030e+07  2.902173e+07  60263.943935
1570  6112.744506  3.321958e+07  2.887157e+07  60452.220918
1571  6111.791472  3.314914e+07  2.869285e+07  60585.218185
1572  5897.338278  3.307860e+07  2.856717e+07  60586.909337
1573  5849.262681  3.313946e+07  2.855823e+07  60606.974473
1574  5849.389398  3.310479e+07  2.854731e+07  60630.105363
1575  5853.111529  3.305654e+07  2.851072e+07  60714.019613

In [642]:
VARMAX_new_states = []
index = -21
for i in range(0,20): 
    previous_state = states.iloc[train_var.index[index:index+1]]
    VARMAX_new_states.append(inverse_transformation_and_state_update(varmax_predictions[i],previous_state,params))
    index += 1
    
VARMAX_new_states = pd.concat(VARMAX_new_states)
VARMAX_new_states

Q             D         R_Rai         R_Eth
1556  6113.017030  3.373426e+07  2.971978e+07  61661.196053
1557  6113.767936  3.369428e+07  2.961953e+07  61701.011749
1558  6111.834120  3.427227e+07  3.018998e+07  60746.861009
1559  6111.676794  3.409202e+07  2.997949e+07  60990.377376
1560  6111.737270  3.412698e+07  2.998928e+07  60974.811810
1561  6111.183947  3.412650e+07  3.002667e+07  60899.716552
1562  6111.416858  3.394189e+07  2.977983e+07  60663.153212
1563  6113.684414  3.379694e+07  2.954039e+07  59907.871761
1564  6111.202474  3.352808e+07  2.943918e+07  60270.855448
1565  6112.309691  3.336299e+07  2.922069e+07  60087.006328
1566  6112.711526  3.336272e+07  2.919058e+07  60163.061682
1567  6112.516972  3.333513e+07  2.913813e+07  60212.264822
1568  6112.598914  3.334894e+07  2.902137e+07  60285.561837
1569  6111.154189  3.334848e+07  2.902173e+07  60263.937977
1570  6111.795664  3.320777e+07  2.887158e+07  60452.390642
1571  6111.617205  3.314979e+07  2.869297e+07  60585.322346
1572  5845.884475  3.306300e+07  2.856728e+07  60586.908513
1573  5845.843064  3.311563e+07  2.855824e+07  60606.965541
1574  5850.375399  3.310543e+07  2.854731e+07  60630.097761
1575  5850.342911  3.305758e+07  2.851071e+07  60714.029589

In [643]:
test_data = states.iloc[test.index[0:20]][['collateral','EthInUniswap','RaiInUniswap','RaiDrawnFromSAFEs']]


In [644]:
test_data

collateral  EthInUniswap  RaiInUniswap  RaiDrawnFromSAFEs
1557   6112.1732  61701.008091  2.961953e+07       3.369223e+07
1558   6112.1732  60746.845297  3.019005e+07       3.427256e+07
1559   6112.1732  60990.371859  2.997948e+07       3.409167e+07
1560   6112.1732  60974.811993  2.998928e+07       3.412678e+07
1561   6112.1732  60899.716831  3.002668e+07       3.412678e+07
1562   6112.1732  60663.156241  2.977982e+07       3.394178e+07
1563   6112.1732  59907.912377  2.954037e+07       3.379492e+07
1564   6112.1732  60270.860599  2.943917e+07       3.352893e+07
1565   6112.1732  60087.012619  2.922068e+07       3.336193e+07
1566   6112.1732  60163.059766  2.919058e+07       3.336193e+07
1567   6112.1732  60212.263728  2.913813e+07       3.333443e+07
1568   6112.1732  60285.559976  2.902136e+07       3.334813e+07
1569   6112.1732  60263.937922  2.902173e+07       3.334893e+07
1570   6112.1732  60452.386579  2.887157e+07       3.320728e+07
1571   6112.1732  60585.321023  2.869296e+07       3.314974e+07
1572   5847.1732  60586.908513  2.856727e+07       3.306341e+07
1573   5847.1732  60606.965445  2.855824e+07       3.311591e+07
1574   5851.6732  60630.097656  2.854731e+07       3.310571e+07
1575   5851.6732  60714.029269  2.851071e+07       3.305789e+07
1576   5851.6732  60767.560015  2.853850e+07       3.308720e+07

## Q - Root Mean Square Error

In [645]:
var_q_rmse = m.sqrt(mean_squared_error(VAR_new_states['Q'], test_data['collateral']))
varmax_q_rmse = m.sqrt(mean_squared_error(VARMAX_new_states['Q'], test_data['collateral']))
if var_q_rmse >= varmax_q_rmse:
    print('VARMAX performs better by {}'.format(varmax_q_rmse-var_q_rmse))
else:
    print('VAR performs better by {}'.format(var_q_rmse-varmax_q_rmse))

VARMAX performs better by -3.4337187137228042


## D - Root Mean Square Error

In [646]:
var_d_rmse = m.sqrt(mean_squared_error(VAR_new_states['D'], test_data['RaiDrawnFromSAFEs']))
varmax_d_rmse = m.sqrt(mean_squared_error(VARMAX_new_states['D'], test_data['RaiDrawnFromSAFEs']))
if var_d_rmse >= varmax_d_rmse:
    print('VARMAX performs better by {}'.format(varmax_d_rmse-var_d_rmse))
else:
    print('VAR performs better by {}'.format(var_d_rmse-varmax_d_rmse))

VAR performs better by -3387.590170986572


## R_Rai - Root Mean Square Error

In [647]:
var_rai_rmse = m.sqrt(mean_squared_error(VAR_new_states['R_Rai'], test_data['RaiInUniswap']))
varmax_rai_rmse = m.sqrt(mean_squared_error(VARMAX_new_states['R_Rai'], test_data['RaiInUniswap']))
if var_rai_rmse >= varmax_rai_rmse:
    print('VARMAX performs better by {}'.format(varmax_rai_rmse-var_rai_rmse))
else:
    print('VAR performs better by {}'.format(var_rai_rmse-varmax_rai_rmse))

VAR performs better by -152.92185181335662


## R_Eth - Root Mean Square Error

In [648]:
var_eth_rmse = m.sqrt(mean_squared_error(VAR_new_states['R_Eth'], test_data['EthInUniswap']))
varmax_eth_rmse = m.sqrt(mean_squared_error(VARMAX_new_states['R_Eth'], test_data['EthInUniswap']))
if var_eth_rmse >= varmax_eth_rmse:
    print('VARMAX performs better by {}'.format(varmax_eth_rmse-var_eth_rmse))
else:
    print('VAR performs better by {}'.format(var_eth_rmse-varmax_eth_rmse))

VARMAX performs better by -0.15262955560427827


In [649]:
print('Aggregated VARMAX RMSE in absolute values:')
print(abs(varmax_eth_rmse) + abs(varmax_rai_rmse) + abs(varmax_d_rmse) + abs(varmax_q_rmse))

print('Aggregated VAR RMSE in absolute values:')
print(abs(var_eth_rmse) + abs(var_rai_rmse) + abs(var_d_rmse) + abs(var_q_rmse))

Aggregated VARMAX RMSE in absolute values:
348061.5987702837
Aggregated VAR RMSE in absolute values:
344524.67309575307


### Exponential weighted moving average test grid

#### Aggregated RMSE values with VARMAX(1) and VAR(6)
VAR lags changed between alphas but for simplicity we held constant

|Alpha   |VARMAX RMSE   | VAR RMSE  | 
|---|---|---|
|0.1   |346,906.48   |356,511.48   |
|0.2  |355,980.48   |353,095.32   |
|0.3  |352,226.20   |350,923.13   |
|0.4  |347,489.97   |348,758.57  |
|0.5  |348,488.02  |347,155.68   |
|0.6  |348,345.53   |346,199.91   |
|0.7  |347,916.76   |345,765.18   |
|**0.8**  |348,061.60   |**345,692.82**   |
|0.9  |349,104.38   |346,223   |

## Conclusion


In this notebook we evaluated the performance of a VARMAX(1) model trained every 20 timesteps an the exogenous signal of redemption price error, against a VAR(15) model retrained everytime step. We decided on using a VAR(15) for the Rai Digital Twin because it performs slightly better, being able to retrain at every prediction timestep. 